### PR TITLE
Optimize one nullable key aggregate performance

### PR DESCRIPTION
**Pull request by liuneng1994**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimize one nullable key aggregate performance

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by liuneng1994** [comment]
This PR optimizes the aggregation performance when the data actually has no NULL or the number of NULL is very small when using the nullable type.
In my use case the performance can be improved by up to 50%.

**Comment by liuneng1994** [comment]
image

**Comment by liuneng1994** [comment]
Does anyone know the reason for the AST fuzzer (tsan) error, it seems to have nothing to do with my code

**Comment by liuneng1994** [comment]
@nickitat The PR check has passed, please review the code when you are free

**Comment by liuneng1994** [comment]
@nickitat Do you have time to review the code?

File: src/Interpreters/Aggregator.h [hunk 1]
```diff
@@ -283,6 +331,37 @@ struct AggregationMethodStringNoCache
     }
 };
 
+template <typename TData>
+struct AggregationMethodStringNullableNoCache
```
Comment (nickitat) [inline src/Interpreters/Aggregator.h:335]: could we reuse existing AggregationMethods code (as well as HashMethods) through a private inheritance for example?

Comment (liuneng1994) [reply src/Interpreters/Aggregator.h]: I added a nullable generic parameter to replace the new struct type

**Comment by liuneng1994** [comment]
@nickitat Please review the code again, I have refactored the code to improve the reusability of the code

File: src/Common/ColumnsHashingImpl.h [hunk 2]
```diff
@@ -188,6 +225,12 @@ class HashMethodBase
             else
                 cache.value = Value();
         }
+        if constexpr (nullable)
+        {
+
+            null_map = &checkAndGetColumn<ColumnNullable>(column)->getNullMapColumn();
+            has_null_data = (null_map->getRatioOfDefaultRows() < 1.0);
```
Comment (nickitat) [inline src/Common/ColumnsHashingImpl.h:232]: let's replace it with a straightforward loop.

File: src/Common/ColumnsHashingImpl.h [hunk 3]
```diff
@@ -172,10 +194,25 @@ class HashMethodBase
         return data.hash(keyHolderGetKey(key_holder));
     }
 
+    ALWAYS_INLINE bool isNullAt(size_t row) const
+    {
+        if constexpr (nullable)
+        {
+            return assert_cast<const ColumnUInt8 &>(*null_map).getData()[row] != 0;
```
Comment (nickitat) [inline src/Common/ColumnsHashingImpl.h:201]: I don't expect it to affect perf, but imo we'd better to do cast only once in ctor

File: src/Common/ColumnsHashingImpl.h (hunk 2 above)
Comment (liuneng1994) [reply src/Common/ColumnsHashingImpl.h]: It doesn't seem to affect performance here, I have deleted the relevant code

File: src/Interpreters/Aggregator.cpp [hunk 4]
```diff
@@ -1912,8 +1937,8 @@ Block Aggregator::insertResultsIntoColumns(PaddedPODArray<AggregateDataPtr> & pl
                     continue;
 
                 auto & final_aggregate_column = out_cols.final_aggregate_columns[i];
-                final_aggregate_column = final_aggregate_column->cloneResized(places.size());
-                columns_data.emplace_back(getColumnData(final_aggregate_column.get()));
+                final_aggregate_column = final_aggregate_column->cloneResized(places.size() + (has_null_key_data ? 1 : 0));
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply src/Interpreters/Aggregator.cpp:1944]: done

File: src/Common/ColumnsHashingImpl.h (hunk 3 above)
Comment (liuneng1994) [reply src/Common/ColumnsHashingImpl.h]: done

File: tests/performance/groupby_onekey_nullable.xml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    <fill_query>insert into t_nullable select ['aa','bb','cc','dd'][number % 10000 + 1], ['aa','bb','cc','dd'][number % 200000 + 1], ['aa','bb','cc','dd'][number % 4 + 1], toInt64(number%10000), toInt64(number%200000), toInt32(number%10000), toInt32(number%200000), toInt16(number%10000), toInt8(number%100), number%6000+1, number%5000+2 from numbers_mt(20000000)</fill_query>
+    <fill_query>insert into t_nullable select ['aa','bb','cc','dd'][number % 10000 + 1], ['aa','bb','cc','dd'][number % 200000 + 1], ['aa','bb','cc','dd'][number % 4 + 1], toInt64(number%10000), toInt64(number%200000), toInt32(number%10000), toInt32(number%200000), toInt16(number%10000), toInt8(number%100), number%6000+1, number%5000+2 from numbers_mt(20000000)</fill_query>
+    <fill_query>OPTIMIZE TABLE t_nullable FINAL</fill_query>
+    <query>select min(i1), min(i2) from t_nullable group by key_string format Null</query>
```
[the remark this answers is not in the document]
Comment (liuneng1994) [reply tests/performance/groupby_onekey_nullable.xml]: done

**Review by nickitat — APPROVED** [review]

**Comment by nickitat** [comment]
hung check failed - #47165

**Pull request merged**